art/attacks/evasion/graphite/utils.py

```diff
@@ -0,0 +1,627 @@
+    crop_off_y: float,
+    pts: np.ndarray,
+    whitebox: bool = False,
+) -> Union[np.ndarray, "torch.Tensor"]:
+    """
+    Compute perspective transform.
+    :param img: Input image.
+    :param angle: Angle to rotate.
+    :param width: Width of image.
+    :param height: Height of image.
+    :param focal: Focal length.
+    :param dist: Distance for transform.
+    :param crop_percent: Percentage for cropping.
+    :param crop_off_x: Cropping x offset.
+    :param crop_off_y: Cropping y offset.
+    :param pts: pts to include in the crop.
+    :param whitebox: Whether this is the white-box version or not.
+    :return: Transformed image.
+    """
+    import torch  # lgtm [py/repeated-import]
+    import cv2  # lgtm [py/repeated-import]
+    from kornia.geometry.transform import warp_perspective
+
+    # img in numpy / cv2 form
+
+    angle = math.radians(angle)
+    x_cam_off = width / 2 - math.sin(angle) * dist
+    z_cam_off = -math.cos(angle) * dist
+    y_cam_off = height / 2
+
+    rot_mat = np.array(
+        [
+            [math.cos(angle), 0, -math.sin(angle), 0],
+            [0, 1, 0, 0],
+            [math.sin(angle), 0, math.cos(angle), 0],
+            [0, 0, 0, 1],
+        ]
+    )
+    c_mat = np.array([[1, 0, 0, -x_cam_off], [0, 1, 0, -y_cam_off], [0, 0, 1, -z_cam_off], [0, 0, 0, 1]])
+
+    rt_mat = np.matmul(rot_mat, c_mat)
+
+    h_mat = np.array(
+        [
+            [focal * rt_mat[0, 0], focal * rt_mat[0, 1], focal * rt_mat[0, 3]],
+            [focal * rt_mat[1, 0], focal * rt_mat[1, 1], focal * rt_mat[1, 3]],
+            [rt_mat[2, 0], rt_mat[2, 1], rt_mat[2, 3]],
+        ]
+    )
+
+    x_off, y_off, crop_size = get_offset_and_crop_size(
+        width, height, h_mat, crop_percent, crop_off_x, crop_off_y, pts, focal / dist
+    )
+
+    affine_mat = np.array([[1, 0, x_off], [0, 1, y_off], [0, 0, 1]])
+    perspective_mat = np.matmul(affine_mat, h_mat)
+
+    if height > width:  # tall and narrow
+        crop_x = crop_size
+        crop_y = int(round(crop_size / width * height))
+    else:  # wide and short or square
+        crop_y = crop_size
+        crop_x = int(round(crop_size / height * width))
+
+    if not whitebox:
+        if not isinstance(img, np.ndarray):
+            raise ValueError("img must be np.ndarray")
+
+        dst = cv2.warpPerspective(img, perspective_mat, (crop_x, crop_y), borderMode=cv2.BORDER_REPLICATE)
+    else:
+        if not isinstance(img, torch.Tensor):
+            raise ValueError("img must be torch tensor")
+
+        dst = warp_perspective(
+            img,
+            torch.from_numpy(perspective_mat).float().to(img.device).unsqueeze(0),
+            (crop_y, crop_x),
+            align_corners=True,
+            padding_mode="border",
+        )
+
+    return dst
+
+
+def get_offset_and_crop_size(
+    width: int,
+    height: int,
+    h_mat: np.ndarray,
+    crop_percent: float,
+    crop_off_x: float,
+    crop_off_y: float,
+    pts: np.ndarray,
```

Comment:
I think on line 355, this arg could be optional, but it is not marked as optional here.